homeassistant/components/humidifier/__init__.py

```diff
@@ -0,0 +1,178 @@
+    ATTR_CURRENT_HUMIDITY,
+    ATTR_HUMIDITY,
+    ATTR_MAX_HUMIDITY,
+    ATTR_MIN_HUMIDITY,
+    ATTR_MODE,
+    DEFAULT_MAX_HUMIDITY,
+    DEFAULT_MIN_HUMIDITY,
+    DEVICE_CLASS_DEHUMIDIFIER,
+    DEVICE_CLASS_HUMIDIFIER,
+    DOMAIN,
+    SERVICE_SET_HUMIDITY,
+    SERVICE_SET_MODE,
+    SUPPORT_MODES,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+SCAN_INTERVAL = timedelta(seconds=60)
+
+DEVICE_CLASSES = [DEVICE_CLASS_HUMIDIFIER, DEVICE_CLASS_DEHUMIDIFIER]
+
+DEVICE_CLASSES_SCHEMA = vol.All(vol.Lower, vol.In(DEVICE_CLASSES))
+
+
+@bind_hass
+def is_on(hass, entity_id):
+    """Return if the humidifier is on based on the statemachine.
+
+    Async friendly.
+    """
+    return hass.states.is_state(entity_id, STATE_ON)
+
+
+async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
+    """Set up humidifier devices."""
+    component = hass.data[DOMAIN] = EntityComponent(
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+    await component.async_setup(config)
+
+    component.async_register_entity_service(SERVICE_TURN_ON, {}, "async_turn_on")
+    component.async_register_entity_service(SERVICE_TURN_OFF, {}, "async_turn_off")
+    component.async_register_entity_service(
+        SERVICE_SET_MODE,
+        make_entity_service_schema({vol.Required(ATTR_MODE): cv.string}),
+        "async_set_mode",
+        [SUPPORT_MODES],
+    )
+    component.async_register_entity_service(
+        SERVICE_SET_HUMIDITY,
+        make_entity_service_schema({vol.Required(ATTR_HUMIDITY): vol.Coerce(int)}),
+        "async_set_humidity",
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    return await hass.data[DOMAIN].async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    return await hass.data[DOMAIN].async_unload_entry(entry)
+
+
+class HumidifierEntity(ToggleEntity):
+    """Representation of a humidifier device."""
+
+    @property
+    def capability_attributes(self) -> Dict[str, Any]:
+        """Return capability attributes."""
+        supported_features = self.supported_features
+        data = {
+            ATTR_MIN_HUMIDITY: self.min_humidity,
+            ATTR_MAX_HUMIDITY: self.max_humidity,
+        }
+
+        if supported_features & SUPPORT_MODES:
+            data[ATTR_AVAILABLE_MODES] = self.available_modes
+
+        return data
+
+    @property
+    def state_attributes(self) -> Dict[str, Any]:
+        """Return the optional state attributes."""
+        supported_features = self.supported_features
+        data = {
+            ATTR_HUMIDITY: self.target_humidity,
+            ATTR_CURRENT_HUMIDITY: self.current_humidity,
+        }
+
+        if supported_features & SUPPORT_MODES:
+            data[ATTR_MODE] = self.mode
+
+        return data
+
+    @property
+    def current_humidity(self) -> Optional[int]:
```

Comment:
I think current humidity should be a separate sensor, right @balloob?

Reply:
Ok, if you say so :)